### ISSUE TITLE
Visual Studio 2015

### ISSUE THREAD

**Issue by cgohlke** [pyopencl/pyopencl#103]
Currently PyOpenCL >= 2015.2 can not be built for CPython 3.5 on Windows because mingwpy does not support Python 3.5.

According to @yuyichao, Visual Studio 2015 support for C++11 should be complete enough to compile PyOpenCL. See https://github.com/pyopencl/pyopencl/commit/d19d94cfb09803059fbd6665283e10617d9e4229#commitcomment-14903890

I am using the following patch to fix many compile errors due to unsupported variadic macro syntax. Does this look reasonable/portable?

```diff
diff --git a/src/c_wrapper/clhelper.h b/src/c_wrapper/clhelper.h
index 0d8f630..2fb6cfd 100644
--- a/src/c_wrapper/clhelper.h
+++ b/src/c_wrapper/clhelper.h
@@ -67,8 +67,8 @@ make_cloutarg(clobj_t *ret, cl_int (CL_API_CALL *release)(typename CLObj::cl_typ
 {
     return _CLObjOutArg<CLObj, T...>(ret, release, name, t1...);
 }
-#define pyopencl_outarg(type, ret, func, args...)               \
-    make_cloutarg<type>(ret, func, #func, ##args)
+#define pyopencl_outarg(type, ret, func, ...)               \
+    make_cloutarg<type>(ret, func, #func, ##__VA_ARGS__)
 
 // {{{ GetInfo helpers
 
@@ -83,8 +83,8 @@ get_vec_info(cl_int (CL_API_CALL *func)(ArgTypes...), const char *name,
     call_guarded(func, name, args..., size_arg(buf), buf_arg(size));
     return buf;
 }
-#define pyopencl_get_vec_info(type, what, args...)                      \
-    get_vec_info<type>(clGet##what##Info, "clGet" #what "Info", args)
+#define pyopencl_get_vec_info(type, what, ...)                      \
+    get_vec_info<type>(clGet##what##Info, "clGet" #what "Info", __VA_ARGS__)
 
 template<typename T>
 PYOPENCL_USE_RESULT static PYOPENCL_INLINE generic_info
@@ -109,8 +109,8 @@ convert_array_info(const char *tname, pyopencl_buf<T> &&_buf)
 
 #define pyopencl_convert_array_info(type, buf)          \
     convert_array_info<type>(#type, buf)
-#define pyopencl_get_array_info(type, what, args...)                    \
-    pyopencl_convert_array_info(type, pyopencl_get_vec_info(type, what, args))
+#define pyopencl_get_array_info(type, what, ...)                    \
+    pyopencl_convert_array_info(type, pyopencl_get_vec_info(type, what, __VA_ARGS__))
 
 template<typename CLObj, typename T>
 PYOPENCL_USE_RESULT static PYOPENCL_INLINE generic_info
@@ -123,9 +123,9 @@ convert_opaque_array_info(T &&buf)
     info.value = buf_to_base<CLObj>(std::forward<T>(buf)).release();
     return info;
 }
-#define pyopencl_get_opaque_array_info(cls, what, args...)  \
+#define pyopencl_get_opaque_array_info(cls, what, ...)  \
     convert_opaque_array_info<cls>(               \
-        pyopencl_get_vec_info(cls::cl_type, what, args))
+        pyopencl_get_vec_info(cls::cl_type, what, __VA_ARGS__))
 
 template<typename CLObj, typename... ArgTypes, typename... ArgTypes2>
 PYOPENCL_USE_RESULT static PYOPENCL_INLINE generic_info
@@ -145,9 +145,9 @@ get_opaque_info(cl_int (CL_API_CALL *func)(ArgTypes...), const char *name,
     }
     return info;
 }
-#define pyopencl_get_opaque_info(clobj, what, args...)              \
+#define pyopencl_get_opaque_info(clobj, what, ...)              \
     get_opaque_info<clobj>(clGet##what##Info,             \
-                                     "clGet" #what "Info", args)
+                                     "clGet" #what "Info", __VA_ARGS__)
 
 template<typename... ArgTypes, typename... ArgTypes2>
 PYOPENCL_USE_RESULT static PYOPENCL_INLINE generic_info
@@ -165,8 +165,8 @@ get_str_info(cl_int (CL_API_CALL *func)(ArgTypes...), const char *name,
     info.value = (void*)param_value.release();
     return info;
 }
-#define pyopencl_get_str_info(what, args...)                            \
-    get_str_info(clGet##what##Info, "clGet" #what "Info", args)
+#define pyopencl_get_str_info(what, ...)                            \
+    get_str_info(clGet##what##Info, "clGet" #what "Info", __VA_ARGS__)
 
 template<typename T, typename... ArgTypes, typename... ArgTypes2>
 PYOPENCL_USE_RESULT static PYOPENCL_INLINE generic_info
@@ -182,9 +182,9 @@ get_int_info(cl_int (CL_API_CALL *func)(ArgTypes...), const char *name,
     info.value = cl_memdup(&value);
     return info;
 }
-#define pyopencl_get_int_info(type, what, args...)                      \
+#define pyopencl_get_int_info(type, what, ...)                      \
     get_int_info<type>(clGet##what##Info, "clGet" #what "Info", \
-                                 #type "*", args)
+                                 #type "*", __VA_ARGS__)
 
 // }}}
 
@@ -200,8 +200,8 @@ convert_obj(cl_int (CL_API_CALL *clRelease)(CLType), const char *name, CLType cl
         throw;
     }
 }
-#define pyopencl_convert_obj(type, func, args...)       \
-    convert_obj<type>(func, #func, args)
+#define pyopencl_convert_obj(type, func, ...)       \
+    convert_obj<type>(func, #func, __VA_ARGS__)
 
 // {{{ extension function pointers
 
diff --git a/src/c_wrapper/error.h b/src/c_wrapper/error.h
index bae569b..3009018 100644
--- a/src/c_wrapper/error.h
+++ b/src/c_wrapper/error.h
@@ -251,8 +251,8 @@ call_guarded(T (CL_API_CALL *func)(ArgTypes...), const char *name, ArgTypes2&&..
     argpack.finish();
     return res;
 }
-#define pyopencl_call_guarded(func, args...)    \
-    call_guarded(func, #func, args)
+#define pyopencl_call_guarded(func, ...)    \
+    call_guarded(func, #func, __VA_ARGS__)
 
 static PYOPENCL_INLINE void
 cleanup_print_error(cl_int status_code, const char *name) noexcept
@@ -275,8 +275,8 @@ call_guarded_cleanup(cl_int (CL_API_CALL *func)(ArgTypes...), const char *name,
         argpack.finish();
     }
 }
-#define pyopencl_call_guarded_cleanup(func, args...)    \
-    call_guarded_cleanup(func, #func, args)
+#define pyopencl_call_guarded_cleanup(func, ...)    \
+    call_guarded_cleanup(func, #func, __VA_ARGS__)
 
 template<typename Func>
 PYOPENCL_USE_RESULT static PYOPENCL_INLINE error*
diff --git a/src/c_wrapper/wrap_constants.cpp b/src/c_wrapper/wrap_constants.cpp
index ae2980e..1a02455 100644
--- a/src/c_wrapper/wrap_constants.cpp
+++ b/src/c_wrapper/wrap_constants.cpp
@@ -10,8 +10,8 @@ void populate_constants(void(*add)(const char*, const char*, int64_t value))
 {
 #define _ADD_ATTR(TYPE, PREFIX, NAME, SUFFIX, ...)      \
       add(TYPE, #NAME, CL_##PREFIX##NAME##SUFFIX)
-#define ADD_ATTR(TYPE, PREFIX, NAME, EXTRA...)  \
-      _ADD_ATTR(TYPE, PREFIX, NAME, EXTRA)
+#define ADD_ATTR(TYPE, PREFIX, NAME, ...)  \
+      _ADD_ATTR(TYPE, PREFIX, NAME, __VA_ARGS__)
 
     // program_kind
     add("program_kind", "UNKNOWN", KND_UNKNOWN);
```
One compile error at [function.h(53)](https://github.com/pyopencl/pyopencl/blob/master/src/c_wrapper/function.h#L53) remains:
```
wrap_cl.cpp
src\c_wrapper\function.h(53): error C2210: 'T' : pack expansions cannot be used as arguments to non-packed parameters in alias templates
src\c_wrapper\function.h(110): note: see reference to class template instantiation 'ArgPack<Convert,Types...>' being compiled
src\c_wrapper\utils.h(226): warning C4800: 'cl_bool': forcing value to bool 'true' or 'false' (performance warning)
```
[This error report](https://connect.microsoft.com/VisualStudio/feedback/details/1341946) could be related, but it should be fixed in the Visual Studio 2015 version I am using.

Any idea how to fix this?

**Comment by cgohlke** [comment]
```
src\c_wrapper\function.h(53): error C2210: 'T' : pack expansions cannot be used as arguments to non-packed parameters in alias templates
```
This looks like a bug in Visual Studio 2015. Visual Studio 2013, LLVM, and mingw do not fail.

**Comment by inducer** [comment]
I've merged your variadic macro tweak in 3b79b2c. I'll take a look to see if we can work around the template issue somehow. @yuyichao, any ideas?

**Comment by yuyichao** [comment]
The variadic macro one makes sense (and I've bite by this a few days ago....)

For the template one, maybe try to include `<tuple>` ?

**Comment by cgohlke** [comment]
Including `<tuple>` does not help. 
I submitted a problem report to Microsoft, using this example, which works on Visual Studio 2013 but fails with 2015:
```c++
#include "stdafx.h"

#include <functional>

template<typename T>
using rm_ref_t = typename std::remove_reference<T>::type;

template<template<typename...> class Convert, typename... Types>
using _ArgPackBase = std::tuple<Convert<rm_ref_t<Types> >...>;

template<template<typename...> class Convert, typename... Types>
class ArgPack : public _ArgPackBase<Convert, Types...> {
public:
    typedef _ArgPackBase<Convert, Types...> tuple_base;
};

int main()
{
    return 0;
}
```

**Comment by yuyichao** [comment]
Thanks. I was going to suggest trying to reduce the code and I'm glad you already did that.

**Comment by yuyichao** [comment]
Might worth trying if inlining some of the aliases helps even if the code would be longer.

**Comment by cgohlke** [comment]
Can you please suggest some inlined code to try?

**Comment by yuyichao** [comment]
Something along the line of

```
template<template<typename...> class Convert, typename... Types>
using _ArgPackBase = std::tuple<Convert<typename std::remove_reference<Types>::type>...>;
```
(if I matched the bracket correctly....)

**Comment by cgohlke** [comment]
That worked. Thank you.
I got PyOpenCL to build without `CL_ENABLE_GL` after a couple of other minor changes that I'll submit later.
The 64-bit binary for Python 3.5 tests OK.

**Comment by inducer** [comment]
@cgohlke To save me some fidgeting, could you share the patch that made it work?

**Comment by cgohlke** [comment]
This patch works with Visual Studio 2015 and CUDA 7.5 when `CL_ENABLE_GL=False`
```diff
diff --git a/src/c_wrapper/buffer.cpp b/src/c_wrapper/buffer.cpp
index 22d7ec1..70e1ff3 100644
--- a/src/c_wrapper/buffer.cpp
+++ b/src/c_wrapper/buffer.cpp
@@ -1,3 +1,4 @@
+#include <algorithm>
 #include "buffer.h"
 #include "context.h"
 #include "command_queue.h"
diff --git a/src/c_wrapper/function.h b/src/c_wrapper/function.h
index 8986991..40525a1 100644
--- a/src/c_wrapper/function.h
+++ b/src/c_wrapper/function.h
@@ -50,7 +50,7 @@ call_tuple(Function &&func, T &&args)
 }

 template<template<typename...> class Convert, typename... Types>
-using _ArgPackBase = std::tuple<Convert<rm_ref_t<Types> >...>;
+using _ArgPackBase = std::tuple<Convert<typename std::remove_reference<Types>::type>...>;

 template<template<typename...> class Convert, typename... Types>
 class ArgPack : public _ArgPackBase<Convert, Types...> {
diff --git a/src/c_wrapper/gl_obj.cpp b/src/c_wrapper/gl_obj.cpp
index 27525ce..ad5cf5b 100644
--- a/src/c_wrapper/gl_obj.cpp
+++ b/src/c_wrapper/gl_obj.cpp
@@ -60,9 +60,9 @@ enqueue_gl_objects(clEnqueueGLObjectFunc func, const char *name,
         mem_objects, num_mem_objects);
     call_guarded(func, name, cq, _mem_objs, _wait_for, event_out(evt));
 }
-#define enqueue_gl_objects(what, args...)                       \
+#define enqueue_gl_objects(what, ...)                       \
     enqueue_gl_objects(clEnqueue##what##GLObjects,              \
-                       "clEnqueue" #what "GLObjects", args)
+                       "clEnqueue" #what "GLObjects", __VA_ARGS__)

 // c wrapper

diff --git a/src/c_wrapper/wrap_cl.h b/src/c_wrapper/wrap_cl.h
index c1feb22..cb8b421 100644
--- a/src/c_wrapper/wrap_cl.h
+++ b/src/c_wrapper/wrap_cl.h
@@ -34,6 +34,7 @@
 #define NOMINMAX
 #include <windows.h>
 #define strdup _strdup
+#define strcasecmp _stricmp
 #else
 #include <unistd.h>
 #define PYOPENCL_HAVE_EVENT_SET_CALLBACK

```
If `CL_ENABLE_GL=True` or using Intel's OpenCL 5.3, compilation still fails:
```
gl_obj.cpp
src\c_wrapper\utils.h(226): warning C4800: 'cl_bool': forcing value to bool 'true' or 'false' (performance warning)
src/c_wrapper/gl_obj.cpp(157): warning C4297: 'get_apple_cgl_share_group': function assumed not to throw an exception but does
src/c_wrapper/gl_obj.cpp(157): note: __declspec(nothrow), throw(), noexcept(true), or noexcept was specified on the function
src\c_wrapper\error.h(233): error C2672: 'CLArgPack<const gl_texture *const ,unsigned int,ArgBuffer<unsigned int,ArgType::SizeOf>,nullptr>::clcall': no matching overloaded function found
src\c_wrapper\clhelper.h(177): note: see reference to function template instantiation 'void call_guarded<const gl_texture*const &,cl_gl_texture_info&,ArgBuffer<unsigned int,ArgType::SizeOf>,nullptr,cl_mem,cl_gl_texture_info,size_t,void*,size_t*>(cl_int (__stdcall *)(cl_mem,cl_gl_texture_info,size_t,void *,size_t *),const char *,const gl_texture *const &,cl_gl_texture_info &,ArgBuffer<unsigned int,ArgType::SizeOf> &&,nullptr &&)' being compiled
src/c_wrapper/gl_obj.cpp(18): note: see reference to function template instantiation 'generic_info get_int_info<GLenum,cl_mem,cl_gl_texture_info,size_t,void*,size_t*,const gl_texture*const ,cl_gl_texture_info&>(cl_int (__stdcall *)(cl_mem,cl_gl_texture_info,size_t,void *,size_t *),const char *,const char *,const gl_texture *const &&,cl_gl_texture_info &)' being compiled
src\c_wrapper\error.h(233): error C2893: Failed to specialize function template 'unknown-type CLArgPack<const gl_texture *const ,unsigned int,ArgBuffer<unsigned int,ArgType::SizeOf>,nullptr>::clcall(Func,const char *)'
src\c_wrapper\error.h(233): note: With the following template arguments:
src\c_wrapper\error.h(233): note: 'Func=cl_int (__stdcall *)(cl_mem,cl_gl_texture_info,size_t,void *,size_t *)'
src\c_wrapper\error.h(233): error C2672: 'CLArgPack<const gl_texture *const ,unsigned int,ArgBuffer<int,ArgType::SizeOf>,nullptr>::clcall': no matching overloaded function found
src\c_wrapper\clhelper.h(177): note: see reference to function template instantiation 'void call_guarded<const gl_texture*const &,cl_gl_texture_info&,ArgBuffer<int,ArgType::SizeOf>,nullptr,cl_mem,cl_gl_texture_info,size_t,void*,size_t*>(cl_int (__stdcall *)(cl_mem,cl_gl_texture_info,size_t,void *,size_t *),const char *,const gl_texture *const &,cl_gl_texture_info &,ArgBuffer<int,ArgType::SizeOf> &&,nullptr &&)' being compiled
src/c_wrapper/gl_obj.cpp(20): note: see reference to function template instantiation 'generic_info get_int_info<GLint,cl_mem,cl_gl_texture_info,size_t,void*,size_t*,const gl_texture*const ,cl_gl_texture_info&>(cl_int (__stdcall *)(cl_mem,cl_gl_texture_info,size_t,void *,size_t *),const char *,const char *,const gl_texture *const &&,cl_gl_texture_info &)' being compiled
src\c_wrapper\error.h(233): error C2893: Failed to specialize function template 'unknown-type CLArgPack<const gl_texture *const ,unsigned int,ArgBuffer<int,ArgType::SizeOf>,nullptr>::clcall(Func,const char *)'
src\c_wrapper\error.h(233): note: With the following template arguments:
src\c_wrapper\error.h(233): note: 'Func=cl_int (__stdcall *)(cl_mem,cl_gl_texture_info,size_t,void *,size_t *)'
```
```
memory_object.cpp
src\c_wrapper\utils.h(226): warning C4800: 'cl_bool': forcing value to bool 'true' or 'false' (performance warning)
src\c_wrapper\clobj.h(108): warning C4267: 'argument': conversion from 'size_t' to 'int', possible loss of data
src/c_wrapper/memory_object.cpp(106): note: see reference to function template instantiation 'pyopencl_buf<CLType> buf_from_class<event,const clobj_t>(T2 *,size_t)' being compiled
        with
        [
            CLType=cl_event,
            T2=const clobj_t
        ]
src\c_wrapper\error.h(233): error C2672: 'CLArgPack<const memory_object *const ,unsigned int,ArgBuffer<unsigned int,ArgType::SizeOf>,nullptr>::clcall': no matching overloaded function found
src\c_wrapper\clhelper.h(177): note: see reference to function template instantiation 'void call_guarded<const memory_object*const &,cl_uint&,ArgBuffer<unsigned int,ArgType::SizeOf>,nullptr,cl_mem,cl_mem_info,size_t,void*,size_t*>(cl_int (__cdecl *)(cl_mem,cl_mem_info,size_t,void *,size_t *),const char *,const memory_object *const &,cl_uint &,ArgBuffer<unsigned int,ArgType::SizeOf> &&,nullptr &&)' being compiled
src/c_wrapper/memory_object.cpp(54): note: see reference to function template instantiation 'generic_info get_int_info<cl_bool,cl_mem,cl_mem_info,size_t,void*,size_t*,const memory_object*const ,cl_uint&>(cl_int (__cdecl *)(cl_mem,cl_mem_info,size_t,void *,size_t *),const char *,const char *,const memory_object *const &&,cl_uint &)' being compiled
src\c_wrapper\error.h(233): error C2893: Failed to specialize function template 'unknown-type CLArgPack<const memory_object *const ,unsigned int,ArgBuffer<unsigned int,ArgType::SizeOf>,nullptr>::clcall(Func,const char *)'
src\c_wrapper\error.h(233): note: With the following template arguments:
src\c_wrapper\error.h(233): note: 'Func=cl_int (__cdecl *)(cl_mem,cl_mem_info,size_t,void *,size_t *)'
```

**Comment by inducer** [comment]
Thanks!

**Comment by inducer** [comment]
@yuyichao: Any clue about the other two issues? The `memory_object.cpp` one is likely more severe, because it indicates an issue with compiling against newer versions of CL.

(Should have mentioned: I merged that patch.)

**Comment by inducer** [comment]
d9b53bb may actually address a few of these. @cgohlke, could you give that a try?

**Comment by cgohlke** [comment]
Thank you. The build with `CL_ENABLE_GL=True` against CUDA SDK now succeeds. 
With Intels OpenCL SDK I get the following error:
```
program.cpp
src\c_wrapper\utils.h(226): warning C4800: 'cl_bool': forcing value to bool 'true' or 'false' (performance warning)
src/c_wrapper/program.cpp(44): warning C4267: 'argument': conversion from 'size_t' to 'int', possible loss of data
src/c_wrapper/program.cpp(52): warning C4267: 'argument': conversion from 'size_t' to 'int', possible loss of data
src/c_wrapper/program.cpp(57): warning C4267: 'argument': conversion from 'size_t' to 'int', possible loss of data
src/c_wrapper/program.cpp(58): warning C4267: 'argument': conversion from 'size_t' to 'int', possible loss of data
src/c_wrapper/program.cpp(59): warning C4267: 'argument': conversion from 'size_t' to 'int', possible loss of data
src/c_wrapper/program.cpp(60): warning C4267: 'argument': conversion from 'size_t' to 'int', possible loss of data
src/c_wrapper/program.cpp(61): warning C4267: 'argument': conversion from 'size_t' to 'int', possible loss of data
src/c_wrapper/program.cpp(250): warning C4267: '=': conversion from 'size_t' to 'uint32_t', possible loss of data
src\c_wrapper\clobj.h(108): warning C4267: 'argument': conversion from 'size_t' to 'int', possible loss of data
src/c_wrapper/program.cpp(108): note: see reference to function template instantiation 'pyopencl_buf<cl_device_id> buf_from_class<device,const clobj_t>(T2 *,size_t)' being compiled
        with
        [
            T2=const clobj_t
        ]
src\c_wrapper\error.h(233): error C2672: 'CLArgPack<const program *const ,const device *,unsigned int,ArgBuffer<unsigned __int64,ArgType::SizeOf>,nullptr>::clcall': no matching overloaded function found
src\c_wrapper\clhelper.h(177): note: see reference to function template instantiation 'void call_guarded<const program*const &,const device*&,cl_program_build_info&,ArgBuffer<unsigned __int64,ArgType::SizeOf>,nullptr,cl_program,cl_device_id,cl_program_build_info,size_t,void*,size_t*>(cl_int (__cdecl *)(cl_program,cl_device_id,cl_program_build_info,size_t,void *,size_t *),const char *,const program *const &,const device *&,cl_program_build_info &,ArgBuffer<unsigned __int64,ArgType::SizeOf> &&,nullptr &&)' being compiled
src/c_wrapper/program.cpp(95): note: see reference to function template instantiation 'generic_info get_int_info<size_t,cl_program,cl_device_id,cl_program_build_info,size_t,void*,size_t*,const program*const ,const device*&,cl_program_build_info&>(cl_int (__cdecl *)(cl_program,cl_device_id,cl_program_build_info,size_t,void *,size_t *),const char *,const char *,const program *const &&,const device *&,cl_program_build_info &)' being compiled
src\c_wrapper\error.h(233): error C2893: Failed to specialize function template 'unknown-type CLArgPack<const program *const ,const device *,unsigned int,ArgBuffer<unsigned __int64,ArgType::SizeOf>,nullptr>::clcall(Func,const char *)'
src\c_wrapper\error.h(233): note: With the following template arguments:
src\c_wrapper\error.h(233): note: 'Func=cl_int (__cdecl *)(cl_program,cl_device_id,cl_program_build_info,size_t,void *,size_t *)'
src\c_wrapper\clobj.h(128): warning C4267: 'argument': conversion from 'size_t' to 'int', possible loss of data
src\c_wrapper\clobj.h(146): note: see reference to function template instantiation 'pyopencl_buf<clbase *> buf_to_base<T,_Val,_Ty>(T2 *,size_t,_Ty &&)' being compiled
        with
        [
            T=kernel,
            _Val=cl_kernel,
            _Ty=bool,
            T2=cl_kernel
        ]
src/c_wrapper/program.cpp(124): note: see reference to function template instantiation 'pyopencl_buf<clbase *> buf_to_base<kernel,pyopencl_buf<cl_kernel>&,bool>(T2,bool &&)' being compiled
        with
        [
            T2=pyopencl_buf<cl_kernel> &
        ]
```

**Comment by inducer** [comment]
448bf7c fixes another missing `PYOPENCL_CL_CASTABLE_THIS` that only gets compiled with CL 2.

**Comment by cgohlke** [comment]
Building with Visual Studio 2015 works now. Thank you. I am closing this issue.
Building with mingwpy for Python <= 3.4 also works but the 64 bit binaries crash.

**Comment by inducer** [comment]
Happy to hear it. I've released 2015.2.4 with these fixes. Thank you very much for your help and patience!
